src/ephy-sync-service.c

```diff
@@ -0,0 +1,1417 @@
+  json = json_node_get_object (json_parser_get_root (parser));
+  alg = json_object_get_string_member (json, "alg");
+
+  if (g_strcmp0 (alg, "RS256") != 0) {
+    g_warning ("Expected algorithm RS256, found %s. Giving up.", alg);
+    goto out;
+  }
+
+  json_parser_load_from_data (parser, payload, -1, NULL);
+  json = json_node_get_object (json_parser_get_root (parser));
+  principal = json_object_get_object_member (json, "principal");
+  email = json_object_get_string_member (principal, "email");
+  uid_email = g_strdup_printf ("%s@%s", self->uid, soup_uri_get_host (uri));
+
+  if (g_strcmp0 (uid_email, email) != 0) {
+    g_warning ("Expected email %s, found %s. Giving up.", uid_email, email);
+    goto out;
+  }
+
+  self->auth_at = json_object_get_int_member (json, "fxa-lastAuthAt");
+  retval = TRUE;
+
+out:
+  g_free (header);
+  g_free (payload);
+  g_free (uid_email);
+  g_strfreev (pieces);
+  g_object_unref (parser);
+  soup_uri_free (uri);
+
+  return retval;
+}
+
+static void
+obtain_storage_credentials_response_cb (SoupSession *session,
+                                        SoupMessage *msg,
+                                        gpointer     user_data)
+{
+  StorageServerRequestAsyncData *data;
+  EphySyncService *service;
+  JsonParser *parser;
+  JsonObject *json;
+  JsonObject *errors;
+  JsonArray *array;
+
+  data = (StorageServerRequestAsyncData *) user_data;
+  service = EPHY_SYNC_SERVICE (data->service);
+
+  parser = json_parser_new ();
+  json_parser_load_from_data (parser, msg->response_body->data, -1, NULL);
+  json = json_node_get_object (json_parser_get_root (parser));
+
+  if (msg->status_code == 200) {
+    service->storage_endpoint = g_strdup (json_object_get_string_member (json, "api_endpoint"));
+    service->storage_credentials_id = g_strdup (json_object_get_string_member (json, "id"));
+    service->storage_credentials_key = g_strdup (json_object_get_string_member (json, "key"));
+    service->storage_credentials_expiry_time = json_object_get_int_member (json, "duration") +
+                                               ephy_sync_utils_current_time_seconds ();
+  } else if (msg->status_code == 401) {
+    array = json_object_get_array_member (json, "errors");
+    errors = json_node_get_object (json_array_get_element (array, 0));
+    g_warning ("Failed to talk to the Token Server: %s: %s",
+               json_object_get_string_member (json, "status"),
+               json_object_get_string_member (errors, "description"));
+    storage_server_request_async_data_free (data);
+    service->locked = FALSE;
+    goto out;
+  } else {
+    g_warning ("Failed to talk to the Token Server, status code %u. "
+               "See https://docs.services.mozilla.com/token/apis.html#error-responses",
+               msg->status_code);
+    storage_server_request_async_data_free (data);
+    service->locked = FALSE;
+    goto out;
+  }
+
+  ephy_sync_service_send_storage_request (service, data);
```

Comment:
This doesn't look like the cleanest way to write the function. Only in the first case above do you ever get to this call, so you can move it up to the first case, and get rid of the goto.